content/blog/2020/07/2020-07-29-git-performance-improvement-phase2.adoc

```diff
@@ -0,0 +1,103 @@
+
+As we all know that performance of a git fetch is highly correlated to the size of the remote repository, our task was to find the difference in performance for the two available git implementations in the link:https://github.com/jenkinsci/git-plugin[Git Plugin], "Git" and "JGit".
+
+Our major finding was that "Git" performs much better than "JGit" when it comes to a large sized repository (>100 MiB). Interestingly, "JGit" performs better than "Git" when size of the repository is less than 100 MiB. 
+
+In this phase, we were successful in coding this derived knowledge from the benchmarks into a new functionality called the 
+GitToolChooser.
+
+
+== GitToolChooser
+
+This class aims to add the functionality of recommending a git implementation on the basis of the size of a repository which has a strong correlation to the performance of git fetch (from Performance Benchmarks).
+
+It utilizes two heuristics to calculate the size:
+
+* Using cached .git dir from Multi-Branch Projects to estimate the size of a repository
+* Providing an extension point which upon implementation can use REST APIs exposed by git service providers like Github, GitLab etc to fetch the size of remote repository.
+
+Will it optimise your Jenkins instance?
+That depends on:
+
+* If you have a multi branch project in your Jenkins instance, the plugin can use that to recommend the optimal git implementation
+* Else if you have a Branch Source Plugin installed in the Jenkins instance, the plugin will recommend a git implementation using REST APIs provided by Github or GitLab respectively.
+
+The architecture and code for this class is at:link:https://github.com/jenkinsci/git-plugin/pull/931[PR-931]
+
+*Note*: This functionality is an upcoming feature in the subsequent Git Plugin release.
+
+=== JMH benchmarks in multiple environments
+
+The benchmarks were being executed on Linux and MacOS machines frequently but there was a need to check if the results gained from those benchmarks will hold true across many platforms which makes the solution (GitToolChooser) platform agnostic.
+
+To test this hypothesis, we performed an experiment:
+
+Running git fetch operation for a 400 MiB sized repository on:
+
+* Windows
+* Free BSD 12
+* ppc64le
+* s390x
+
+The result of running this experiment is given below:
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-multiple-platforms.png[Performance on multiple platforms]
+
+Observations:
+
+* ppc64le and s390x are able to run the operation in almost half the time it takes for the Windows or Free BSD 12 machine. This behavior may be attributed to the increased computational power of those machines.
+* The difference in performance between Git and JGit remains constant across all platforms which is a positive sign for the GitToolChooser as its recommendation would be consistent across multiple devices and operating systems.
```

Comment:
May want to apply consistent formatting as above if you'll format tool names for git/JGit